packages/grid_client/src/clients/tf-grid/contracts.ts

```diff
@@ -301,13 +331,216 @@ class TFContracts extends Contracts {
+      return USD.div(tftPrice);
+    } catch (error) {
+      throw new GridClientError(`Failed to convert to mTFT due: ${error}`);
+    }
+  }
+
+  /**
+   * Calculate total IPV4 and total overdraft on all node contracts on rented node.
+   * @description In node contracts on rented nodes we need to have both the overdraft to be added to rent contract overdraft.
+   * please note that the unbilled NU amount is added to the total overdraft exactly as the in the other contract types.
+   * the IPV4 cost to add to the estimated cost of the rent contract.
+   */
+  private async getContractsCostOnRentedNode(nodeId: number, proxy: GridProxyClient) {
+    const contracts = await proxy.contracts.list({ nodeId, state: [ContractState.GracePeriod], numberOfPublicIps: 1 });
+
+    const ipPrice = (await this.client.pricingPolicies.get({ id: 1 })).ipu.value / 10 ** 7;
+    const BillingInformationPromises = contracts.data.reduce((acc: Promise<BillingInformation>[], contract) => {
+      acc.push(this.client.contracts.getContractBillingInformationByID(contract.contract_id));
+      return acc;
+    }, []);
+    const overDraftPromises = contracts.data.reduce((acc: Promise<ContractPaymentState>[], contract) => {
+      acc.push(this.client.contracts.getContractPaymentState(contract.contract_id));
+      return acc;
+    }, []);
+
+    const overDraftResult = await Promise.all(overDraftPromises);
+    const billingInfoResult = await Promise.all(BillingInformationPromises);
+
+    const totalOverDraft = overDraftResult.reduce(
+      (acc: Decimal, paymentState) => acc.add(paymentState.additionalOverdraft).add(paymentState.standardOverdraft),
+      new Decimal(0),
+    );
+    const totalNUCost = billingInfoResult.reduce((acc: number, billingInfo) => acc + billingInfo.amountUnbilled, 0);
+    const totalIPCost = (ipPrice * (contracts.data.length || 0)) / 1e7;
+    return {
+      //return ip cost per month
+      totalIpCost: totalIPCost * 24 * 30,
+      totalOverdraft: totalOverDraft.add(totalNUCost),
+    };
+  }
+  /**
+   * A function to calculate the cost of the resources.
+   * @param {CalcResourcesPrice} options please not that the passed {Resources} should be in bytes.
+   * the function will convert them to GB and calculate the cost
+   * @returns { Object } contains the `dedicatedPrice` and the `sharedPrice` both are numbers
+   */
+  private async calcResourcesPrice(options: CalcResourcesPrice) {
+    const { cru, hru, sru, mru, calculator, ipv4, certified } = options;
+
+    const mruInGB = mru / Math.pow(1024, 3);
+    const sruInGB = sru / Math.pow(1024, 3);
+    const hruInGB = hru / Math.pow(1024, 3);
+
+    const res = await calculator.calculateWithMyBalance({
+      cru,
+      mru: mruInGB,
+      sru: sruInGB,
+      hru: hruInGB,
+      certified,
+      ipv4u: !!ipv4,
+    });
+
+    return {
+      dedicatedPrice: res.dedicatedPrice,
+      sharedPrice: res.sharedPrice,
+    };
+  }
```

Comment:
this is a private function to reduce redundant code and make the conversion from bytes to GB inside it instead of doing it multiple times, should be moved to the calculator module ?

Reply:
yes, it would be a better place